### PR TITLE
DOC: Implementing redirect system, and adding user_guide redirects

### LINKED ISSUE #24705: DOC: Redirect moved pages
Following  up the discussion in https://github.com/pandas-dev/pandas/issues/24451#issuecomment-450862635 and addressing the concerns of merging the PRs of #24499.

The problem is that we've moved many documentation pages (all the API pages, `release.rst`...) we'd like to move many more (almost every page). And we've got two options if we've been discussion so far are:
- Break all the links in search engines, stack overflow, blog posts... linking to the old locations.
- Add more complexity to the build simiar to https://github.com/pandas-dev/pandas/blob/master/doc/source/conf.py#L246 (which IMO is already quite complex, and just works for specific cases)

After thinking more on it, I think we can build a simpler redirect system than what we have, and keep redirects while reducing the complexity of the build.

What I'd do is create a file (e.g. `doc/source/redirects.csv`) containing the urls of the the old and new url (a line could look like `io,user_guide/io` or `generated/pandas.DataFrame.to_parquet,api/reference/DataFrame/pandas.DataFrame.to_parquet`)

Then, in `doc/source/conf.py` we open and iterate the file, and create a simple html in the old location with a link to the new. May be later we can use a nice template and a javascript that automatically redirects, but I'd keep it simple to start with.

If a page is splitted, we can have more than two columns in the csv (e.g. `io,user_guide/io_plain,user_guide/io_binary`), and show the two links in the redirect page.

I'd also move the current API redirect system to this, and we'd simplify `conf.py`. 

@jorisvandenbossche @TomAugspurger does this sound good?

### PR DESCRIPTION
**Pull request by datapythonista**
- [X] closes #24705
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: doc/make.py [hunk 1]
```diff
@@ -139,6 +143,72 @@ def _open_browser(self, single_doc_html):
...
+
+                try:
+                    title = self._get_page_title(row[1])
+                except Exception:
+                    # the file can be an ipynb and not an rst, or docutils
+                    # may not be able to read the rst because it has some
+                    # sphinx specific stuff
+                    title = 'this page'
+
+                with open(path, 'w') as redirects_fd:
```
Comment (jreback) [inline doc/make.py]: why wouldn't you open with before the loop?

Comment (datapythonista) [reply doc/make.py]: There are two files, the csv file with the mapping of redirects, which is opened before the loop, and this one, that redirects the old page to the new, which is a different file in every iteration of the loop. May be could find a name less ambiguous, but I think this is correct.

Am I understanding what you mean?

File: doc/make.py [hunk 2]
```diff
@@ -15,15 +15,19 @@
...
+import docutils
+import docutils.parsers.rst
 
 
 DOC_PATH = os.path.dirname(os.path.abspath(__file__))
 SOURCE_PATH = os.path.join(DOC_PATH, 'source')
 BUILD_PATH = os.path.join(DOC_PATH, 'build')
-BUILD_DIRS = ['doctrees', 'html', 'latex', 'plots', '_static', '_templates']
+REDIRECTS_FILE = os.path.join(DOC_PATH, 'redirects.csv')
+BASE_REDIRECT_URL = 'https://pandas.pydata.org/pandas-docs/stable/'
```
[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply doc/make.py]: I think the current redirect system uses relative link in href, which seems to work OK

**Review by jorisvandenbossche — COMMENTED** [review]
On the issue you mentioned to integrate this with (or remove) the current redirect system we have in conf.py, but that wasn't feasible in the end?

**Comment by jorisvandenbossche** [comment]
Also wondering: is there a technical reason you went with a custom approach in make.py instead of using the current logic in conf.py using sphinx's `additional_html_pages` (but then generating the old/new links from the file instead of the current list in the file) ?

**Comment by datapythonista** [comment]
I prefer to remove the previous redirects in a follow up, after this is working. I think it should be feasible. This new system won't introspect classes to redirect all the methods at runtime, which will make the list of redirects longer, but I still think it's better this way.

`additional_html_pages` could be used too, but I like this approach. I think it's better keeping the system separate from sphinx (sphinx is already too complex to mix other things there). And the context for all the additional html pages is the same, which means that we need to add all the redirects to that context, and not just pass the required data. After adding the title of the new location, that would be two somehow big dictionaries, and some trickiness in the templates.

Regarding the absolute redirects, the current/previous system only redirects to the same directory. In out case that's not the case. And our docs live in a subdirectory (`/pandas-docs/stable/`), which means that we can't use absolute url without domain (`/user_guide/missing_data/`). This leaves us with two options, using relative urls from the old to the new (e.g. `../../reference/api`), or using absolute urls. I think the relative are too complex. And knowing that there are plans to use `pandas.io`, I think an absolute url is better, as we could redirect from the old website to the new, if that make sense (not sure if it does, there are different options there).

May be we could not always use the stable documentation, and redirect the master documentation to the master. But I don't think those redirects are actually useful in master, and I wouldn't make things more complex for that. If having the master docs redirecting to stable causes confusion, I'd simply add an option `--no-redirects` (or the opposite) to `make.py`, and not generate them for master.

**Comment by TomAugspurger** [comment]
I'm not sure I follow the relative vs. absolute link discussion. If I go to `<...>/advanced.html`, then why doesn't using `./user_guide/advanced` for the redirect? Then this will work for the dev docs as well (useful for testing).

> And knowing that there are plans to use pandas.io, I think an absolute url is better, as we could redirect from the old website to the new, if that make sense

Ideally that'll be done at the DNS layer, before the webserver actually sees the request.

**Comment by datapythonista** [comment]
About redirecting the user_guide, you're right, is as simple as removing the url. But for the api it'd be `../api/generated`, and I was thinking on possibly even more complex redirects, but on a second thought, I don't think they are likely to happen.

I made the redirects relative, I guess we're all happy with it.

**Review by TomAugspurger — APPROVED** [review]
Looks good, thanks. I'd say merge when you're ready @datapythonista.

**Comment by datapythonista** [comment]
I finally added here the redirects for the 3 restructing PRs (getting started, user guide, development).

I think that'll make things easier, I'll merge those 3 PRs first, and then this one, when everything is green.

**Comment by datapythonista** [comment]
@TomAugspurger I merged the 3 PRs with the restructuring, and added all the redirects here. Restarted the CI, to make sure the doc build finish on green (it doesn't if the redirects overwrite any file).

Feel free to merge on green, if you're happy.

**Review by TomAugspurger — APPROVED** [review]

**Comment by TomAugspurger** [comment]
Thanks! https://travis-ci.org/pandas-dev/pandas/jobs/483537121 is the builder from master.

**Comment by TomAugspurger** [comment]
That build finished. So visiting, e.g. http://pandas-docs.github.io/pandas-docs-travis/contributing.html will redirect to http://pandas-docs.github.io/pandas-docs-travis/development/contributing.html

**Pull request merged**